toolbox/githubctl/main.go

```diff
@@ -79,167 +80,97 @@ func fastForward(repo, baseBranch, refSHA *string) error {
-		cmd := fmt.Sprintf("./install/updateVersion.sh")
-		// Auth
-		cmd += fmt.Sprintf(" -c %s -A %s", hubCommaTag, debianURL)
-		// Mixer
-		cmd += fmt.Sprintf(" -x %s", hubCommaTag)
-		// Pilot
-		cmd += fmt.Sprintf(" -p %s -i %s -P %s", hubCommaTag, istioctlURL, debianURL)
-		// Proxy
-		cmd += fmt.Sprintf(" -r %s -E %s", releaseTag, debianURL)
-		_, err = u.Shell(cmd)
-		return err
+	for _, job := range postSubmitJobs {
+		cache[job] = make(map[string]bool)
+		runNumber, err := prowAccessor.GetLatestRun(job)
+		if err != nil {
+			log.Fatalf("failed to get latest run number of %s: %v", job, err)
+		}
+		// download the most recent prow results up to maxRunDepth
+		for i := 0; i < *maxRunDepth; i++ {
+			tasks <- task{job, runNumber}
+			runNumber--
+		}
 	}
-	releaseBranch := "Istio_Release_" + releaseTag
-	body := "Update istio.Version"
-	prTitle := releasePRTtilePrefix + body
-	_, err = githubClnt.CreatePRUpdateRepo(releaseBranch, *baseBranch, istioRepo, prTitle, body, edit)
-	return err
+	close(tasks)
+	wg.Wait()
+	return cache
 }
 
-// CreateIstioReleaseUploadArtifacts creates a release on istio from the refSHA provided and uploads dependent artifacts
-func CreateIstioReleaseUploadArtifacts() error {
-	u.AssertNotEmpty("ref_sha", refSHA)
+func getLatestGreenSHA() (string, error) {
+	u.AssertNotEmpty("repo", repo)
 	u.AssertNotEmpty("base_branch", baseBranch)
-	u.AssertNotEmpty("next_release", nextRelease)
-	releaseTag, err := getReleaseTag()
-	if releaseTag == *nextRelease {
-		return fmt.Errorf("next_release should be greater than the current release")
-	}
+	u.AssertPositive("max_commit_depth", maxCommitDepth)
+	u.AssertPositive("max_run_depth", maxRunDepth)
+	results := preprocessProwResults()
+	sha, err := githubClnt.GetHeadCommitSHA(*repo, *baseBranch)
 	if err != nil {
-		return err
-	}
-	releaseBranch := "finalizeRelease-" + releaseTag
-	prBody := fmt.Sprintf("Finalize release %s on istio", releaseTag)
-	updateVersion := func() error {
-		log.Printf("Updating downloadIstioCandidate.sh with latest release")
-		return u.UpdateKeyValueInFile(
-			downloadScript, "ISTIO_VERSION", fmt.Sprintf("${ISTIO_VERSION:-%s}", releaseTag))
+		log.Fatalf("failed to get the head commit sha of %s/%s: %v", *repo, *baseBranch, err)
 	}
-	edit := func() error {
-		if _, err = u.Shell(
-			"./release/create_release_archives.sh -d " + releaseBaseDir); err != nil {
-			return err
-		}
-		archiveDir := releaseBaseDir + "/archives"
-		if err = githubClnt.CreateReleaseUploadArchives(
-			istioRepo, releaseTag, *refSHA, archiveDir); err != nil {
-			return err
+	for i := 0; i < *maxCommitDepth; i++ {
+		log.Printf("Checking if [%s] passed all checks. %d commits before HEAD", sha, i)
+		allChecksPassed := true
+		for _, job := range postSubmitJobs {
+			passed, keyExists := results[job][sha]
+			if !keyExists {
+				log.Printf("Results unknown in local cache for [%s] at [%s], treat the test as failed", job, sha)
+				passed = false
```

Comment:
you dont need this line, when key does not exist you get default value which is false in this case

Reply:
done